README.md
# zipSniper

Read the directory of a ZIP Archive remotely. 

* Supports ZIP64
* Support for HTTP and SOCKS proxies in development
```
Extracts a file list within a zip archive remotely

Usage: zipSniper.exe [OPTIONS] --path <url>

Options:
  -p, --path <url>
  -c, --comment-buffer <BYTES>  Number of bytes to pull from the end of the file.
                                The EOCD checksum needs to land in this data chunk.
                                ZIP:0x06054b50  ZIP64:0x06054b50
                                 [default: 56]
  -o, --output-file <FILE>      Sets an optional output file
  -l, --log-level <LEVEL>       Sets the log level (error, warn, info, debug, trace) [default: info]
      --proxy <PROXY_URL>       Sets the proxy to route HTTP requests through
  -h, --help                    Print help
  -V, --version                 Print version
```
UPDATE COMING SOON


Changelog.md
# Changelog

## Versio 0.3
* Major refactor - significant reduction in third party libraries
* Added input validation for proxy server protocol

## Version 0.2
* Add support for http, https and socks5h proxies


## Version 0.1
* Initial Release

src/lib.rs
pub mod zip;

use reqwest;
use zip::*;

pub struct Client {
    remote_path: String,
    comment_buffer: u64,
    client: reqwest::Client,
}

impl Client {
    pub async fn new(
        remote_path: String,
        comment_buffer: u64,
        proxy_address: Option<String>,
    ) -> Client {
        Client {
            remote_path,
            comment_buffer,
            client: Client::build_client(proxy_address),
        }
    }

    fn build_client(proxy_address: Option<String>) -> reqwest::Client {
        proxy_address
            .map(|proxy_address| {
                let proxy = if proxy_address.starts_with("http://") {
                    reqwest::Proxy::http(proxy_address)
                } else if proxy_address.starts_with("https://") {
                    reqwest::Proxy::https(proxy_address)
                } else {
                    reqwest::Proxy::all(proxy_address)
                }
                .expect("Invalid proxy");

                reqwest::Client::builder()
                    .proxy(proxy)
                    .build()
                    .expect("Error building reqwest client")
            })
            .unwrap_or_else(reqwest::Client::new)
    }

    pub async fn download(&self, cd: zip::cd::Cd) {}

    pub async fn build_zip(&self) -> Zip {
        let build_eocd = |this: &Self| async move {
            println!("[WARNING] - The communicating webserver might not support the Range Header.");
            println!("If the tool hangs, it is likely trying to download the full ZIP\n");
            let response = self
                .client
                .get(&self.remote_path)
                .header("Range", format!("bytes=-{}", &self.comment_buffer))
                .send()
                .await
                .expect("Error sending EOCD chunk request");

            let data = response.bytes().await.expect("Error pulling EOCD chunk");

            for (index, window) in data.windows(zip::eocd::SIGNATURE_32.len()).enumerate() {
                if window == zip::eocd::SIGNATURE_32 || window == zip::eocd::SIGNATURE_64 {
                    return eocd::Eocd::from(&data[index..]);
                }
            }
            panic!("EOCD Signature not found. Increase comment buffer");
        };

        let build_central_directory = |this: &Self, cd_offset: u64, cd_size: u64| async move {
            let mut cd_list: Vec<zip::cd::Cd> = Vec::new();
            let response = self
                .client
                .get(&self.remote_path)
                .header(
                    "Range",
                    format!("bytes={}-{}", cd_offset, cd_size + cd_offset),
                )
                .send()
                .await
                .expect("Error sending CD chunk request");

            let data = response.bytes().await.expect("Error pulling CD chunk");
            for (index, window) in data.windows(zip::cd::CD_SIGNATURE.len()).enumerate() {
                if window == zip::cd::CD_SIGNATURE {
                    cd_list.push(cd::Cd::from(&data[index..]));
                }
            }
            return cd_list;
        };

        let eocd = build_eocd(self).await;
        let cd_list =
            build_central_directory(self, eocd.cd_start_offset(), eocd.size_of_cd()).await;
        Zip::build(eocd, cd_list)
    }
}


src/main.rs
use clap::Parser;
use zipSniper;

#[derive(Parser)]
#[command(author, version, about, long_about = None)]
struct Cli {
    // Mandatory path of the remote ZIP file.
    #[arg(value_name = "PATH", help = "Remote path of the ZIP archive.")]
    remote_path: String,

    // Comment buffer
    #[arg(
        short,
        long,
        value_name = "BYTES",
        default_value_t = 56,
        help = "See https://w.wiki/73sX for details."
    )]
    comment_buffer: u64,

    // Proxy address
    #[arg(
        short,
        long,
        value_name = "PATH",
        value_parser = proxy_protocol_validation,
        help = "Address of a proxy (http, https, socks5h)."
    )]
    proxy_address: Option<String>,

    #[arg(
        short,
        long,
        value_name = "SUB-STRING",
        help = "Output filter by filename substring"
    )]
    filter: Option<String>,

    #[arg(
        short,
        long,
        value_name = "SUB-STRING",
        help = "Download filter by filename substring"
    )]
    download: Option<String>,

    #[arg(
        short,
        long,
        default_value_t = false,
        help = "Print all information (could be noisy)"
    )]
    all_information: bool,
}

#[tokio::main]
async fn main() {
    let cli = Cli::parse();
    let client = zipSniper::Client::new(cli.remote_path, cli.comment_buffer, cli.proxy_address);

    let zip = client.await.build_zip().await;
    let mut cd_list = zip.cd_list;

    // If a filter is passed, remove any irrelevant CD records.
    if cli.filter.is_some() {
        let filter = cli.filter.unwrap();
        cd_list = cd_list
            .into_iter()
            .filter(|cd| cd.file_name().contains(&filter))
            .collect();
    }

    for cd in cd_list {
        if cli.all_information == true {
            println!("{}", cd)
        } else {
            println!("{}", cd.file_name())
        }
    }
}

fn proxy_protocol_validation(proxy_address: &str) -> Result<String, String> {
    if proxy_address.starts_with("http://")
        || proxy_address.starts_with("https://")
        || proxy_address.starts_with("socks5h://")
    {
        Ok(String::from(proxy_address))
    } else {
        Err(format!(
            "Invalid Proxy Scheme: {}. Check --help",
            proxy_address
        ))
    }
}


src/zip/cd.rs
use std::fmt;

pub const CD_SIGNATURE: [u8; 4] = [0x50, 0x4b, 0x01, 0x02];
const EXTRA_FIELD_SIGNATURE: [u8; 2] = [0x01, 0x00];

#[derive(Debug)]
pub enum WordSize {
    Bit32,
    Bit64,
}

#[derive(Debug)]
pub struct Cd {
    bytes: Box<[u8]>,
    signature: [u8; 4],
}

pub struct ExtraField<'a> {
    bytes: Box<&'a [u8]>,
    signature: [u8; 2],
}

#[allow(dead_code)]
impl fmt::Display for Cd {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "### Central Directory File: {} ###\n", self.file_name());
        write!(f, "Memory Offset \t Attribute Name \t Value\n");
        match self.verify_wordsize() {
            WordSize::Bit32 => {
                write!(
                    f,
                    "00000004 \t Version Used \t\t {:02X}{:02X}\n",
                    self.bytes[5], self.bytes[4],
                )
                .expect("Error printing out EOCD");
                write!(
                    f,
                    "00000006 \t Min Version Needed \t {:02X}{:02X}\n",
                    self.bytes[7], self.bytes[6],
                )
                .expect("Error printing out EOCD");
                write!(
                    f,
                    "00000008 \t General Purpose Flag \t {:02X}{:02X}\n",
                    self.bytes[9], self.bytes[8],
                )
                .expect("Error printing out EOCD");
                write!(
                    f,
                    "0000000A \t Compression Method \t {:02X}{:02X}\n",
                    self.bytes[11], self.bytes[10],
                )
                .expect("Error printing out EOCD");
                write!(
                    f,
                    "0000000C \t Modification Time \t {:02X}{:02X}\n",
                    self.bytes[13], self.bytes[12],
                )
                .expect("Error printing out EOCD");
                write!(
                    f,
                    "0000000E \t Modification Date \t {:02X}{:02X}\n",
                    self.bytes[15], self.bytes[14],
                )
                .expect("Error printing out EOCD");
                write!(
                    f,
                    "0000000F \t CRC-32 Hash \t\t {:02X}{:02X}{:02X}{:02X}\n",
                    self.bytes[19], self.bytes[18], self.bytes[17], self.bytes[16],
                )
                .expect("Error printing out EOCD");
                write!(
                    f,
                    "00000014 \t Compressed Size \t {:02X}{:02X}{:02X}{:02X}\n",
                    self.bytes[23], self.bytes[22], self.bytes[21], self.bytes[20],
                )
                .expect("Error printing out EOCD");
                write!(
                    f,
                    "00000018 \t Uncompressed Size \t {:02X}{:02X}{:02X}{:02X}\n",
                    self.bytes[27], self.bytes[26], self.bytes[25], self.bytes[24],
                )
                .expect("Error printing out EOCD");
                write!(
                    f,
                    "0000001C \t File Name Length \t {:02X}{:02X}\n",
                    self.bytes[29], self.bytes[28],
                )
                .expect("Error printing out EOCD");
                write!(
                    f,
                    "0000001E \t Extra Field Length \t {:02X}{:02X}\n",
                    self.bytes[31], self.bytes[30],
                )
                .expect("Error printing out EOCD");
                write!(
                    f,
                    "00000020 \t File Comment Length \t {:02X}{:02X}\n",
                    self.bytes[33], self.bytes[32],
                )
                .expect("Error printing out EOCD");
                write!(
                    f,
                    "00000022 \t Disk Number \t\t {:02X}{:02X}\n",
                    self.bytes[35], self.bytes[34],
                )
                .expect("Error printing out EOCD");
                write!(
                    f,
                    "00000024 \t Internal Attributes \t {:02X}{:02X}\n",
                    self.bytes[37], self.bytes[36],
                )
                .expect("Error printing out EOCD");
                write!(
                    f,
                    "00000026 \t External Attributes \t {:02X}{:02X}{:02X}{:02X}\n",
                    self.bytes[41], self.bytes[40], self.bytes[39], self.bytes[38],
                )
                .expect("Error printing out EOCD");
                write!(
                    f,
                    "0000002A \t Local Header Offset \t {:02X}{:02X}{:02X}{:02X}\n",
                    self.bytes[45], self.bytes[44], self.bytes[43], self.bytes[42],
                )
                .expect("Error printing out EOCD");
                write!(f, "0000002E \t File Name \t\t '{}'\n", self.file_name())
                    .expect("Error printing out EOCD");
                write!(
                    f,
                    "{:08X} \t File Comment \t\t '{}'\n",
                    0x2a + self.file_name_length() + self.extra_field_length(),
                    self.file_comment(),
                )
            }
            WordSize::Bit64 => {
                write!(f, "Print Statement not implemented yet.")
            }
        }
    }
}

#[allow(dead_code)]
impl Cd {
    pub fn from(bytes: &[u8]) -> Cd {
        Cd {
            bytes: Box::from(bytes),
            signature: [bytes[0], bytes[1], bytes[2], bytes[3]],
        }
    }

    pub fn verify_wordsize(&self) -> WordSize {
        let _compressed_size = [
            self.bytes[20],
            self.bytes[21],
            self.bytes[22],
            self.bytes[23],
        ];
        let _uncompressed_size = [
            self.bytes[24],
            self.bytes[25],
            self.bytes[26],
            self.bytes[27],
        ];
        let _relative_offset = [
            self.bytes[42],
            self.bytes[43],
            self.bytes[44],
            self.bytes[45],
        ];
        let _disk_number = [self.bytes[34], self.bytes[35]];

        if _compressed_size == [0xff, 0xff, 0xff, 0xff]
            || _uncompressed_size == [0xff, 0xff, 0xff, 0xff]
            || _relative_offset == [0xff, 0xff, 0xff, 0xff]
            || _disk_number == [0xff, 0xff]
        {
            WordSize::Bit64
        } else {
            WordSize::Bit32
        }
    }

    pub fn version_made_by(&self) -> u16 {
        u16::from_le_bytes([self.bytes[4], self.bytes[5]])
    }

    pub fn version_needed_to_extract(&self) -> u16 {
        u16::from_le_bytes([self.bytes[6], self.bytes[7]])
    }

    pub fn general_purpose_flag(&self) -> u16 {
        u16::from_le_bytes([self.bytes[8], self.bytes[9]])
    }

    pub fn compression_method(&self) -> u16 {
        u16::from_le_bytes([self.bytes[10], self.bytes[11]])
    }

    pub fn last_modification_time(&self) -> u16 {
        u16::from_le_bytes([self.bytes[12], self.bytes[13]])
    }

    pub fn last_modification_date(&self) -> u16 {
        u16::from_le_bytes([self.bytes[13], self.bytes[14]])
    }

    pub fn crc32_hash(&self) -> u32 {
        u32::from_le_bytes([
            self.bytes[15],
            self.bytes[16],
            self.bytes[17],
            self.bytes[18],
        ])
    }

    pub fn compressed_size(&self) -> u64 {
        u64::from_le_bytes(match self.verify_wordsize() {
            WordSize::Bit32 => [
                self.bytes[20],
                self.bytes[21],
                self.bytes[22],
                self.bytes[23],
                0,
                0,
                0,
                0,
            ],
            WordSize::Bit64 => {
                let offset = 46 + self.file_name_length() as usize;
                [
                    self.bytes[12 + offset],
                    self.bytes[13 + offset],
                    self.bytes[14 + offset],
                    self.bytes[15 + offset],
                    self.bytes[16 + offset],
                    self.bytes[17 + offset],
                    self.bytes[18 + offset],
                    self.bytes[19 + offset],
                ]
            }
        })
    }

    pub fn uncompressed_size(&self) -> u64 {
        u64::from_le_bytes(match self.verify_wordsize() {
            WordSize::Bit32 => [
                self.bytes[24],
                self.bytes[25],
                self.bytes[26],
                self.bytes[27],
                0,
                0,
                0,
                0,
            ],
            WordSize::Bit64 => {
                let offset = 46 + self.file_name_length() as usize;
                [
                    self.bytes[4 + offset],
                    self.bytes[5 + offset],
                    self.bytes[6 + offset],
                    self.bytes[7 + offset],
                    self.bytes[8 + offset],
                    self.bytes[9 + offset],
                    self.bytes[10 + offset],
                    self.bytes[11 + offset],
                ]
            }
        })
    }

    pub fn file_name_length(&self) -> u16 {
        u16::from_le_bytes([self.bytes[28], self.bytes[29]])
    }

    pub fn extra_field_length(&self) -> u16 {
        u16::from_le_bytes([self.bytes[30], self.bytes[31]])
    }

    pub fn file_comment_length(&self) -> u16 {
        u16::from_le_bytes([self.bytes[32], self.bytes[33]])
    }

    pub fn disk_number_where_file_starts(&self) -> u32 {
        u32::from_le_bytes(match self.verify_wordsize() {
            WordSize::Bit32 => [self.bytes[34], self.bytes[35], 0, 0],
            WordSize::Bit64 => {
                let offset = 46 + self.file_name_length() as usize;
                [
                    self.bytes[28 + offset],
                    self.bytes[29 + offset],
                    self.bytes[30 + offset],
                    self.bytes[31 + offset],
                ]
            }
        })
    }

    pub fn internal_file_attributes(&self) -> u16 {
        u16::from_le_bytes([self.bytes[36], self.bytes[37]])
    }

    pub fn external_file_attributes(&self) -> u32 {
        u32::from_le_bytes([
            self.bytes[38],
            self.bytes[39],
            self.bytes[40],
            self.bytes[41],
        ])
    }

    pub fn relative_offset_local_header(&self) -> u64 {
        u64::from_le_bytes(match self.verify_wordsize() {
            WordSize::Bit32 => [
                self.bytes[42],
                self.bytes[43],
                self.bytes[44],
                self.bytes[45],
                0,
                0,
                0,
                0,
            ],
            WordSize::Bit64 => {
                let offset = 46 + self.file_name_length() as usize;
                [
                    self.bytes[20 + offset],
                    self.bytes[21 + offset],
                    self.bytes[22 + offset],
                    self.bytes[23 + offset],
                    self.bytes[24 + offset],
                    self.bytes[25 + offset],
                    self.bytes[26 + offset],
                    self.bytes[27 + offset],
                ]
            }
        })
    }

    pub fn file_name(&self) -> String {
        let last_byte = 46 + self.file_name_length() as usize;
        String::from_utf8(self.bytes[46..last_byte].to_vec()).expect("Invalid UTF-8")
    }

    pub fn extra_field(&self) -> ExtraField {
        let first_byte = 46 + self.file_name_length() as usize;
        let last_byte = first_byte + self.extra_field_length() as usize;
        let signature = [self.bytes[first_byte], self.bytes[first_byte + 1]];
        if signature != EXTRA_FIELD_SIGNATURE {
            panic!("Invalid Extra Field Signature");
        }
        ExtraField {
            bytes: Box::new(&self.bytes[first_byte..last_byte]),
            signature: signature,
        }
    }

    pub fn file_comment(&self) -> String {
        let mut first_byte = (46 + self.file_name_length()) as usize;
        let mut last_byte = (46 + self.file_name_length() + self.file_comment_length()) as usize;
        match self.verify_wordsize() {
            WordSize::Bit32 => {}
            WordSize::Bit64 => {
                first_byte += self.extra_field_length() as usize;
                last_byte += self.extra_field_length() as usize;
            }
        }
        String::from_utf8(self.bytes[first_byte..last_byte].to_vec()).expect("Invalid UTF-8")
    }
}


zipSniper.py
import requests
import argparse
import struct
import uuid
import os
from tqdm import tqdm

class eocd_struct_64:

    def __init__(self, bytestream):

        self.bytestream = bytestream
        unpacked_stream = struct.unpack("<LQHHLLQQQQ", self.bytestream[:56])
        self.size_of_eocd64 = unpacked_stream[1]
        self.version_made_by = unpacked_stream[2]
        self.min_version_needed = unpacked_stream[3]
        self.disk_number = unpacked_stream[4]
        self.disk_where_cd_starts = unpacked_stream[5]
        self.number_of_cd_records_on_disk = unpacked_stream[6]
        self.total_number_of_cd_records = unpacked_stream[7]
        self.cd_size = unpacked_stream[8]
        self.cd_start_offset = unpacked_stream[9]

class eocd_struct:

    def __init__(self, bytestream):

        self.bytestream = bytestream
        unpacked_stream = struct.unpack("<LHHHHIIH", self.bytestream[:22])
        self.disk_number = unpacked_stream[1]
        self.disk_where_cd_starts = unpacked_stream[2]
        self.number_of_cd_records_on_disk = unpacked_stream[3]
        self.total_number_of_cd_records = unpacked_stream[4]
        self.cd_size = unpacked_stream[5]
        self.cd_start_offset = unpacked_stream[6]

class cd_struct:

    def __init__(self, bytestream):

        self.bytestream = bytestream
        unpacked_stream = struct.unpack("<IHHHHHHIIIHHHHHII", self.bytestream[:46])
        self.version_made_by = unpacked_stream[1]
        self.min_version_needed = unpacked_stream[2]
        self.general_purpose_flag = unpacked_stream[3]
        self.compresion_method = unpacked_stream[4]
        self.last_modification_time = unpacked_stream[5]
        self.last_modification_date = unpacked_stream[6]
        self.crc32_uncompressed_checksum = unpacked_stream[7]
        self.compressed_size = unpacked_stream[8]
        self.uncompressed_size = unpacked_stream[9]
        self.file_name_length = unpacked_stream[10]
        self.extra_field_length = unpacked_stream[11]
        self.file_comment_length = unpacked_stream[12]
        self.disk_number_where_file_starts = unpacked_stream[13]
        self.internal_file_attributes = unpacked_stream[14]
        self.external_file_attributes = unpacked_stream[15]
        self.relative_offset = unpacked_stream[16]
        self.file_name = bytestream[46:46+self.file_name_length]

class zipSniper():

    def __init__(self, remote_path: str, comment_buffer: int, http_proxy, https_proxy):

        self.remote_path = remote_path
        self.directory_list = list()
        self.eocd_blob = None
        self.proxies = dict()
        if http_proxy is not None:
            self.proxies["http"] = http_proxy
        if https_proxy is not None:
            self.proxies["https"] = https_proxy
        if comment_buffer < 0:
            self.comment_buffer = comment_buffer * -1
        else:
            self.comment_buffer = comment_buffer
        self.zip_size = self._zip_size()
        self.is64 = False
        self._eocd_blob()
        self._cd_blob()
        self._sanity_check()

    def _zip_size(self):

        response = requests.head(
            self.remote_path,
            allow_redirects=True,
            proxies=self.proxies
        )

        result = int(response.headers["Content-Length"])
        print(f"ZIP Size: {result}")
        return result

    def _eocd_blob(self):

        response = requests.get(
            self.remote_path,
            allow_redirects=True,
            headers={"Range": f"bytes=-{self.comment_buffer}"},
            proxies=self.proxies
        )

        eocd_sig_64 = b'\x50\x4b\x06\x06'
        eocd_sig = b'\x50\x4b\x05\x06'

        binary_blob = bytearray(response.content)
        for position, byte in enumerate(binary_blob):
            if hex(byte) == "0x50":
                if eocd_sig == binary_blob[position:position+4]:
                    self.is64 = False
                elif eocd_sig_64 == binary_blob[position:position+4]:
                    self.is64 = True
                else:
                    continue
                print(f"Is Zip64: {self.is64}")
                print(f"EOCD Blob offset: {hex(self.zip_size - position)}")
                if self.is64:
                    self.eocd_blob = eocd_struct_64(binary_blob[position:self.zip_size])
                else:
                    self.eocd_blob = eocd_struct(binary_blob[position:self.zip_size])
                return

        raise MissingEocdSig(self.comment_buffer)
    
    def _cd_blob(self):

        cd_offset = self.eocd_blob.cd_start_offset
        cd_size = self.eocd_blob.cd_size

        response = requests.get(
            self.remote_path,
            allow_redirects=True,
            headers={"Range": f"bytes={cd_offset}-{cd_offset + cd_size - 1}"},
            proxies=self.proxies,
            stream=True
        )

        block_size = 1024
        progress_bar = tqdm(total=(cd_size - 1), unit="iB", unit_scale=True)
        binary_blob = bytearray()

        filename = str(uuid.uuid4())
        with open(filename, "wb") as file:
            for d in response.iter_content(block_size):
                progress_bar.update(len(d))
                file.write(d)

        progress_bar.close()

        cd_sig = b'\x50\x4b\x01\x02'
        
        binary_blob = open(filename, "rb").read()
        for position, byte in enumerate(binary_blob):
            if hex(byte) == "0x50":
                if cd_sig == binary_blob[position:position+4]:
                    cd_data = cd_struct(binary_blob[position:-1])
                    self.directory_list.append(cd_data.file_name.decode("utf-8"))
        os.remove(filename)

    def _sanity_check(self):

        if len(self.directory_list) != self.eocd_blob.total_number_of_cd_records:
            raise IncompleteDirectoryList(len(self.directory_list), self.eocd_blob.total_number_of_cd_records)

                
class MissingEocdSig(Exception):

    def __init__(self, comment_buffer, *args):
        self.comment_buffer = comment_buffer
        self.message = f"Comment Buffer of {comment_buffer} is too small!"
        super().__init__(self.message, *args)


class IncompleteDirectoryList(Exception):

    def __init__(self, correct_size, found_size, *args):
        self.correct_size = correct_size
        self.found_size = found_size
        self.message = f"EOCD Reports {correct_size} CD Records. Only {found_size} found!"
        super().__init__(self.message, *args)


if __name__ == "__main__":

    parser = argparse.ArgumentParser(
        prog="zipSniper",
        description="Pulls the directory structure of a zip file remotely",
    )
    parser.add_argument("url", type=str, help="Remote path of the zip archive")
    parser.add_argument("-c", 
                        "--comment_buffer",
                        dest="comment_buffer", 
                        type=int, 
                        default=1024,
                        metavar="int",
                        help="The size of the comment in EOCD is unknown. So to ensure we can find the EOCDs position in the ZIP archive, we specify a random buffer (Default: 1024 Bytes)"
                    )
    parser.add_argument("--http_proxy", dest="http_proxy", type=str, default=None, metavar="url")
    parser.add_argument("--https_proxy", dest="https_proxy", type=str, default=None, metavar="url")
    parser.add_argument("-O", "--output_file", dest="output_file", type=str, default=None)
    args = parser.parse_args()

    sniper = zipSniper(args.url, args.comment_buffer, args.http_proxy, args.https_proxy)
    if args.output_file is not None:
        with open(args.output_file, "w") as file:
            for d in sniper.directory_list:
                file.write(d + "\n")
    else:
        print(sniper.directory_list)
       
    #work to be done
